matrix-appservice-discord: 1.0.0 -> 3.1.0

pacien opened the pull request:
###### Description of changes

This updates the package for `matrix-appservice-discord` to the latest version available.
Changelog: https://github.com/matrix-org/matrix-appservice-discord/releases/tag/v3.0.0

Tested and confirmed to work on my own deployment.


###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

georgyo commented:
@pacien are you still working on this?

pacien commented:
Quoting George Shammas (2022-08-31 18:53:37)
> @pacien are you still working on this?

No, I kind of gave up on this after getting stuck dealing with some
dependency:
https://github.com/NixOS/nixpkgs/pull/186316/files#diff-898cf569590a55c53c2c2284784d59a326b51c2c783ea354ddb917d114398e8fR31-R35

Any help appreciated :)

pacien commented:
Quoting ash lea (2022-09-05 15:04:10)
> @pacien i got this working on my fork:
> https://github.com/ashkitten/nixpkgs/commit/dbd1118fedce28ac39897445ffa819d9cdcaa422
>
> currently just downloads the aarch64 lib (my server is aarch64), so
> it'll need to be fixed up to support multiple architectures

I would prefer building from the sources instead of pulling a prebuilt
binary, but if we're going that latter route, let's not forget to add
`meta.sourceProvenance = [ lib.sourceTypes.binaryNativeCode ]`.

ashkitten commented:
yes, we can build from source, but i think it would have to be a separate derivation that just builds the `.node` lib

pacien committed:
matrix-appservice-discord: 1.0.0 -> 3.1.0

- This also takes into account the transfer of the project and repository
  to the matrix-org organisation on GitHub.

- The Node version assertion has been removed because our `pkgs.nodejs`
  already points to version 18.

- This replaces the `.nix` lock file (generated with `yarn2nix`) with a
  hash-pinned fixed-output derivation, which is simpler to work with.

- This replaces the old "generate.sh" script with one that updates all the
  version info and hashes (package.json, source version, source hash,
  and Yarn dependencies hash).
  This is inspired by the update script used for the
  "matrix-appservice-slack" package.

- This switches to srcOnly for node sources.

- Added pkgs.matrix-sdk-crypto-nodejs new required native dependency.

pacien commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -11,20 +19,33 @@ let
+  };
 
   pkgConfig = {
+    "@matrix-org/matrix-sdk-crypto-nodejs" = {
+      # Building this native Rust dependency requires the `napi` tool from
+      # `napi-rs`, which isn't available in nixpkgs at the time of writing.
+      # Because appservice-discord does not really make use of this library
+      # anyway, we can snip it off to neutralise loading errors.
+      postInstall = ''
+        truncate -s0 index.js
+      '';
+    };
+
```
All tests are passing with this, but I didn't try running the program on my
server. It would be nice if someone could try actually running the service with
this to confirm that this isn't breaking anything.

pacien replied:
Works fine in my deployment.

pacien commented:
@ofborg eval

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1197

SuperSandro2000 commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -1,6 +1,14 @@
+, nodejs
+}:
+
+assert lib.versionAtLeast nodejs.version "16.0.0";
 
```
```suggestion
```

nodejs points to 18 already

SuperSandro2000 commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -11,20 +19,33 @@ let
 
-  # when updating, run `./generate.sh <git release tag>`
-  version = "1.0.0";
+  # when changing version, make sure to update the local copy of `package.json`
```
```suggestion
```
Please add an update script instead

SuperSandro2000 commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -56,7 +77,9 @@ in mkYarnPackage rec {
     # admin tools wrappers
     for toolPath in $OUT_JS_DIR/tools/*; do
-      makeWrapper '${nodejs}/bin/node' "$out/bin/${pname}-$(basename $toolPath .js)" \
+      makeWrapper \
+        '${nodejs}/bin/node' \
+        "$out/bin/${pname}-$(basename $toolPath .js)" \
```
Please revert

SuperSandro2000 commented on pkgs/servers/matrix-appservice-discord/generate.sh:
```diff
@@ -1,17 +0,0 @@
-#!/usr/bin/env nix-shell
-#!nix-shell -I nixpkgs=../../../ -i bash -p wget yarn2nix
-
-set -euo pipefail
-
-if [ "$#" -ne 1 ] || [[ "$1" == -* ]]; then
-  echo "Regenerates the Yarn dependency lock files for the matrix-appservice-discord package."
-  echo "Usage: $0 <git release tag>"
-  exit 1
-fi
-
-SRC_REPO="https://raw.githubusercontent.com/Half-Shot/matrix-appservice-discord/$1"
-
-wget "$SRC_REPO/package.json" -O package.json
-wget "$SRC_REPO/yarn.lock" -O yarn.lock
-yarn2nix --lockfile=yarn.lock > yarn-dependencies.nix
-rm yarn.lock
```
Please keep that

pacien commented:
Quoting Sandro (2022-09-17 22:19:04)
> > +assert lib.versionAtLeast nodejs.version "16.0.0";
> 
> nodejs points to 18 already

Outdated check removed :+1:

> > +      makeWrapper \
> +        '${nodejs}/bin/node' \
> +        "$out/bin/${pname}-$(basename $toolPath .js)" \
> 
> Please revert

Revert the long lines wrapping? I find it clearer and easier to work
with once wrapped, for both diffing and reading.

> > -#!/usr/bin/env nix-shell
> > -#!nix-shell -I nixpkgs=../../../ -i bash -p wget yarn2nix
> > [...]
> 
> Please keep that

> Please add an update script instead

Update script restored (and simplified due to not needing to run
`yarn2nix` manually anymore).

SuperSandro2000 commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -12,19 +18,32 @@ in mkYarnPackage rec {
-  yarnNix = ./yarn-dependencies.nix;
+  offlineCache = fetchYarnDeps {
+    yarnLock = "${src}/yarn.lock";
+    sha256 = "sha256-UPhibxSqadu69OfY0FheXxRWK7H2SU+6awzUp8kGo4I=";
```
This isn't automatically updated at the moment, right?

pacien commented:
Quoting Sandro (2022-09-22 15:16:49)
> @SuperSandro2000 commented on this pull request.
> > +      makeWrapper \
> +        '${nodejs}/bin/node' \
> +        "$out/bin/${pname}-$(basename $toolPath .js)" \
> 
> ```suggestion
>       makeWrapper '${nodejs}/bin/node' \
>         "$out/bin/${pname}-$(basename $toolPath .js)" \
> ```

Amended.

pacien commented:
Quoting Sandro (2022-09-22 15:17:40)
> @SuperSandro2000 commented on this pull request.
> -  yarnNix = ./yarn-dependencies.nix;
> +  offlineCache = fetchYarnDeps {
> +    yarnLock = "${src}/yarn.lock";
> +    sha256 = "sha256-UPhibxSqadu69OfY0FheXxRWK7H2SU+6awzUp8kGo4I=";
> 
> This isn't automatically updated at the moment, right?

I don't think it's automatically updated.
Running `./generate.sh` is also still necessary to update the
package.json.

SuperSandro2000 commented:
> I don't think it's automatically updated. Running `./generate.sh` is also still necessary to update the package.json.

The update script should handle everything or nothing because otherwise this will be forgotten.

pacien commented:
Quoting Sandro (2022-10-03 21:56:13)
> The update script should handle everything or nothing because
> otherwise this will be forgotten.

Should we remove "generate.sh" then?

It's essentially just running "wget" to download the latest
"package.json".

pacien commented:
Quoting Sandro (2022-10-06 11:52:23)
> The update script should handle that, too.

That's now handled by the update script :+1:

I also rebased the branch and included the changes from #192266.

All feedbacks should have been taken into account now.

pacien commented:
Quoting Sandro (2022-09-17 22:19:04)
> > +assert lib.versionAtLeast nodejs.version "16.0.0";
> 
> nodejs points to 18 already

Outdated check removed :+1:

> > +      makeWrapper \
> +        '${nodejs}/bin/node' \
> +        "$out/bin/${pname}-$(basename $toolPath .js)" \
> 
> Please revert

Revert the long lines wrapping? I find it clearer and easier to work
with once wrapped, for both diffing and reading.

> > -#!/usr/bin/env nix-shell
> > -#!nix-shell -I nixpkgs=../../../ -i bash -p wget yarn2nix
> > [...]
> 
> Please keep that

> Please add an update script instead

Update script restored (and simplified due to not needing to run
`yarn2nix` manually anymore).

winterqt commented on pkgs/servers/matrix-appservice-discord/update.sh:
```diff
@@ -0,0 +1,42 @@
+curl -O "$src/package.json"
+cat > pin.json << EOF
+{
+  "version": "$version",
```
The `v` should be stripped from the tag here, as it causes an incorrect `version` and `rev`.

pacien commented:
Quoting Winter (2022-10-18 13:39:48)
> @winterqt requested changes on this pull request.
> +  "version": "$version",
> 
> The `v` should be stripped from the tag here, as it causes an
> incorrect `version` and `rev`.

Good catch; That's now fixed :+1:

winterqt commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -1,35 +1,57 @@
 
 let
+  pin = lib.importJSON ./pin.json;
   nodeSources = runCommand "node-sources" {} ''
     tar --no-same-owner --no-same-permissions -xf ${nodejs.src}
     mv node-* $out
   '';
```
Prefer `srcOnly nodejs` here.

winterqt commented on pkgs/servers/matrix-appservice-discord/default.nix:
```diff
@@ -1,35 +1,57 @@
 
   pkgConfig = {
+    "@matrix-org/matrix-sdk-crypto-nodejs" = {
+      # Building this native Rust dependency requires the `napi` tool from
+      # `napi-rs`, which isn't available in nixpkgs at the time of writing.
+      # Because appservice-discord does not really make use of this library
+      # anyway, we can snip it off to neutralise loading errors.
```
This isn't true any more -- do you think it's worth copying the library, or no (since it's not used)?

winterqt replied:
You unresolved this without responding -- mind sharing your opinion?

pacien replied:
My email reply landed outside of the thread it seems:
https://github.com/NixOS/nixpkgs/pull/186316#issuecomment-1282450133

winterqt commented:
Also: I think some of these commits can be squashed. For example, I don't think we need a large auto-generated dependencies diff and then a switch to fetchYarnDeps (deleting that file) right after. If you disagree, though, happy to discuss.

pacien commented:
Quoting Winter (2022-10-18 15:09:26)
> @winterqt requested changes on this pull request.
> >    nodeSources = runCommand "node-sources" {} ''
>      tar --no-same-owner --no-same-permissions -xf ${nodejs.src}
>      mv node-* $out
>    '';
> 
> Prefer `srcOnly nodejs` here.

Replaced. That's indeed nicer :+1:

> > +      # Building this native Rust dependency requires the `napi` tool from
> +      # `napi-rs`, which isn't available in nixpkgs at the time of writing.
> +      # Because appservice-discord does not really make use of this library
> +      # anyway, we can snip it off to neutralise loading errors.
> 
> This isn't true any more -- do you think it's worth copying the
> library, or no (since it's not used)?

`pkgs.matrix-sdk-crypto-nodejs` does not seem to build anymore on master
as of today. I propose that we include this lib later when it will be
actually needed by this bridge and fixed.

pacien commented:
Quoting Winter (2022-10-18 15:12:13)
> Also: I think some of these commits can be squashed. For example, I
> don't think we need a large auto-generated dependencies diff and then
> a switch to fetchYarnDeps (deleting that file) right after. If you
> disagree, though, happy to discuss.

I squashed the fixups :+1:

I prefer to keep the rest in separate commits in order to show the
individual steps to make the review easier.

winterqt commented:
> I prefer to keep the rest in separate commits in order to show the individual steps to make the review easier.

I think the fetchYarnDeps diff is small enough (minus the deletion of the file of course) that it can be merged into it.

🤷‍♀️ Not sure what's best here, I'm going to ask around for opinions.

pacien commented:
Rebased to solve merge conflict.
I believe there's nothing left to do here.

pacien commented:
Updated package from 3.0.0 to 3.1.0, which was released in the meantime.

georgyo commented:
This all looks good to me, tested the current branch head and it is working great as well.

@SuperSandro2000 Any blockers on merging this finally?

winterqt reviewed (changes requested):
After [discussing with some others on Matrix](https://matrix.to/#/!kjdutkOsheZdjqYmqp%3Anixos.org/%244xnK6C7WKMe0n5WstzuKCN7CqijMBlksLDPyb-YRdJQ):

1. Please squash the 3.1.0 and 3.0.0 bump -- we generally don't like having intermediate bumps.
2. I and others am of the opinion that the `fetchYarnDeps` migration should be squashed into the previous commit -- there's no reason for the intermediate dependency file modifications.
3. Maybe drop the "wrap lines" commit/squash it somewhere? Not sure, but it was mentioned by someone so I figured I'd put it here.

Thanks, and sorry this is taking so long.

pacien commented:
Squashed all because it's too complicated otherwise.

winterqt commented:
> `pkgs.matrix-sdk-crypto-nodejs` does not seem to build anymore on master as of today. I propose that we include this lib later when it will be actually needed by this bridge and fixed.

Now that it does build, do you want to add it?

pacien commented:
Quoting Winter (2022-11-10 02:59:49)
> > `pkgs.matrix-sdk-crypto-nodejs` does not seem to build anymore on
> > master as of today. I propose that we include this lib later when it
> > will be actually needed by this bridge and fixed.
> 
> Now that it does build, do you want to add it?

Added and amended :+1:

winterqt committed:
matrix-appservice-discord: run pre/post hooks for overridden phases

winterqt reviewed (approved):
Thanks!
